code/datums/elements/immerse.dm

```diff
@@ -0,0 +1,387 @@
+		return
+
+	var/atom/movable/buckled
+	if(isliving(movable))
+		var/mob/living/living_mob = movable
+		RegisterSignal(living_mob, COMSIG_LIVING_SET_BUCKLED, PROC_REF(on_set_buckled))
+		buckled = living_mob.buckled
+
+	try_immerse(movable, buckled)
+	RegisterSignal(movable, COMSIG_QDELETING, PROC_REF(on_movable_qdel))
+	LAZYADD(attached_turfs_and_movables[source], movable)
+	ADD_TRAIT(movable, TRAIT_IMMERSED, ELEMENT_TRAIT(src))
+
+/datum/element/immerse/proc/on_movable_qdel(atom/movable/source)
+	SIGNAL_HANDLER
+	remove_from_element(source.loc, source)
+
+/**
+ * The main proc, which adds a visual overlay to the movable that has entered the turf to make it look immersed.
+ * It's kind of iffy but basically, we want the overlay to cover as much area as needed to
+ * avoid the movable's icon from spilling horizontally or below.
+ * Also, while these visual overlays are mainly cached movables, for certain movables, such as living mobs,
+ * we want them to have their own unique vis overlay with additional signals registered.
+ * This allows the vis overlay to look more or less unchanged while its owner is spinning or resting
+ * without otherwise affecting other movables with identical overlays.
+ */
+/datum/element/immerse/proc/add_immerse_overlay(atom/movable/movable)
+	var/icon/movable_icon = icon(movable.icon)
+	var/width = movable_icon.Width() || world.icon_size
+	var/height = movable_icon.Height() || world.icon_size
+
+	var/is_below_water = movable.layer < WATER_LEVEL_LAYER ? "underwater-" : ""
+
+	var/atom/movable/immerse_overlay/vis_overlay = generated_visual_overlays["[is_below_water][width]x[height]"]
+
+	if(!vis_overlay) //create the overlay if not already done.
+		vis_overlay = new(null, src)
+
+		/**
+		 * vis contents spin around the center of the icon of their vis locs
+		 * but since we want the appearance to stay where it should be,
+		 * we have to counteract this one.
+		 */
+		var/extra_width = (width - world.icon_size) * 0.5
+		var/extra_height = (height - world.icon_size) * 0.5
+		var/mutable_appearance/overlay_appearance = new()
+		var/icon/immerse_icon = generated_immerse_icons["[icon]-[icon_state]-[mask_icon]"]
+		var/last_i = width/world.icon_size
+		for(var/i in -1 to last_i)
+			var/mutable_appearance/underwater = mutable_appearance(icon, icon_state)
+			underwater.pixel_x = world.icon_size * i - extra_width
+			underwater.pixel_y = -world.icon_size - extra_height
+			overlay_appearance.overlays += underwater
+
+			var/mutable_appearance/water_level = is_below_water ? underwater : mutable_appearance(immerse_icon)
+			water_level.pixel_x = world.icon_size * i - extra_width
+			water_level.pixel_y = -extra_height
+			overlay_appearance.overlays += water_level
+
+
+		vis_overlay.color = color
+		vis_overlay.alpha = alpha
+		vis_overlay.overlays = list(overlay_appearance)
+
+		vis_overlay.extra_width = extra_width
+		vis_overlay.extra_height = extra_height
+		vis_overlay.overlay_appearance = overlay_appearance
```

Comment:
I think it would be worth moving this into its own proc, for readability and performance tracking